### DISCUSSION TITLE
Void event type for connection limits and allow-block list behaviours.

### DISCUSSION THREAD

**Discussion by shamil-gadelshin** [libp2p/rust-libp2p#3977] [General]
Libp2p v0.51.3

When we use connection limits or allow-block list behaviours along with `#[behaviour(out_event = "Event")]` directive:
```
#[derive(NetworkBehaviour)]
#[behaviour(out_event = "Event")]
#[behaviour(event_process = false)]
pub(crate) struct Behavior {
..
    pub(crate) connection_limits: ConnectionLimitsBehaviour,
    pub(crate) block_list: BlockListBehaviour,
}
```
we need to specify the `Void` type for the event:
```
#[derive(Debug, From)]
pub(crate) enum Event {
..
    VoidEventStub(Void),
}
```
Otherwise, we get an error: the trait `std::convert::From<void::Void>` is not implemented for `behavior::Event`.

It's confusing for a couple of reasons:
- Void type is not exported from libp2p and must be added separately
- we need to specify an additional event stub for both behaviours.

Is there a better option to work with events for connection limits or allow-block list behaviours?

**Comment by thomaseizinger** [comment]
> * Void type is not exported from libp2p and must be added separately

`void` is > 1.0 so I think it is safe to appear in public APIs and can easily be depended on separately.

> Is there a better option to work with events for connection limits or allow-block list behaviours?

Not to my knowledge. Ideally, we would use the `never` (`!`) here but that isn't stable yet.

Does it work if you let the derive macro auto-generate the enum? You'll have two `Void` variants then but those are easily handled in a `match` using `void::unreachable`: https://docs.rs/void/1.0.2/void/fn.unreachable.html

**Comment by shamil-gadelshin** [comment]
> void is > 1.0 so I think it is safe to appear in public APIs and can easily be depended on separately.

I see. You seem to prefer a separate third-party crate for such dependencies rather than reexported types.

> Does it work if you let the derive macro auto-generate the enum?

We tend to have more control over events` names.

It works either way. Thanks for the answers.

**Reply by thomaseizinger** [reply]
> > void is > 1.0 so I think it is safe to appear in public APIs and can easily be depended on separately.
> 
> I see. You seem to prefer a separate third-party crate for such dependencies rather than reexported types.

Re-exporting types makes dependency upgrades easier because you can directly access the type in the correct version.

At some point though, this has to stop. We don't re-export `tokio` or `async-std` either but they still appear in our public APIs. `futures`, `either` and `serde` are other examples.

Latest once a crate signals stability (> 1.0), I think it is safe to stop re-exporting it and have the user declare a dependency themselves. Otherwise, you end up with lots of your dependencies all re-exporting the same version. That makes the import path redundant which I'd argue is confusing.

Simply put, once a crate is stable, the _need_ for re-exporting goes away. It no longer solves a problem and I like to not have code around that doesn't solve any problems :)

**Reply by shamil-gadelshin** [reply]
It makes sense. Thanks for the clarification!

**Comment by nazar-pc** [comment]
Right now `Void` is a public API of the crate. Re-exporting `void` makes little sense, but `Void` must be IMHO. It can change from re-exported to concrete type in the future if necessary, making transition easier for users. It is far less ubiquitous than `serde` and the likes.

**Reply by thomaseizinger** [reply]
We can replace it with `Infallible`, then you have to write the `match e { }` yourself instead of using `void::unreachable(e)`. Would you prefer that?

**Reply by mxinden** [reply]
In my eyes the type safety gained through `void::unreachable` is worth the dependency.

**Reply by thomaseizinger** [reply]
The other thing we could try is  https://docs.rs/never-say-never/latest/never_say_never/. I believe that is considered directly exhaustive without matching on it.

That would remove the need for calling a fn like `void::unreachable`.

Also, it implements `From` for any type: https://doc.rust-lang.org/src/core/convert/mod.rs.html#751

This should remove the need for declaring `From` impls if you declare your own `OutEvent` using the derive.

**Reply by thomaseizinger** [reply]
@nazar-pc @shamil-gadelshin Would either one of you be open to experiment with using the `never-say-never` crate where we are currently using `void`?

**Reply by nazar-pc** [reply]
My concern was never that `void` crate is used, but rather that it is used in one specific situation, but the type is not re-exported. It would be very low priority for me to do anything about it.

Maybe skipping the event type for behaviors that don't have any events would be even better, my understanding that is why `Void` is used in the first place.

**Reply by mxinden** [reply]
> My concern was never that void crate is used, but rather that it is used in one specific situation, but the type is not re-exported. It would be very low priority for me to do anything about it.

My understanding with `never-say-never`, given that it doesn't need the `match`, there is no need for downstream users to explicitly name it, and thus no need for rust-libp2p to re-export it. @thomaseizinger please correct me in case I am missing something.

**Reply by shamil-gadelshin** [reply]
I could spend some free time on an experiment with `never-say-never` and one of the behaviors. We could briefly touch this on the next rust-libp2p maintainers call.

**Reply by thomaseizinger** [reply]
> > My concern was never that void crate is used, but rather that it is used in one specific situation, but the type is not re-exported. It would be very low priority for me to do anything about it.
> 
> My understanding with `never-say-never`, given that it doesn't need the `match`, there is no need for downstream users to explicitly name it, and thus no need for rust-libp2p to re-export it. @thomaseizinger please correct me in case I am missing something.

Yes exactly, that is the hypothesis.

**Reply by thomaseizinger** [reply]
> I could spend some free time on an experiment with `never-say-never` and one of the behaviors. We could briefly touch this on the next rust-libp2p maintainers call. 

That would be great! The next one will be on the upcoming Tuesday.

**Reply by shamil-gadelshin** [reply]
Following the maintainers call: https://github.com/shamil-gadelshin/rust-libp2p/commit/c971dc67d9c8f79dcbae96db7d61268d69eb703d   @thomaseizinger

**Comment by thomaseizinger** [comment]
I just realized how we can fix your original issue @shamil-gadelshin !

We can extend the derive-macro to always generate the following impl if `out_event` is specified.

```rust
impl From<#derive_prelude::Void> for #out_event {
	fn from(v: #derive_prelude::Void) -> Self {
        void::unreachable(v)
	}
}
```

We will also have to add `Void` to our `derive_prelude` module in `libp2p-swarm`. This satisfies the trait-bounds and allows you to omit all `From` impls for `Void`.

Are you open to send a PR?

**Reply by shamil-gadelshin** [reply]
Let me try it.

**Reply by shamil-gadelshin** [reply]
@thomaseizinger  Here is my attempt: https://github.com/shamil-gadelshin/rust-libp2p/commit/d6bb609568855a79ce7ce801ddcc81134687d8a6

The straightforward solution doesn't cover all the cases: `custom_out_event_no_type_parameters` test fails to compile.

**Reply by thomaseizinger** [reply]
Can you open a PR please?

**Reply by shamil-gadelshin** [reply]
> Can you open a PR please?

Sure, I'll do that when I have a working version.

**Reply by shamil-gadelshin** [reply]
On the second look, it seems that a problem with the test lies deeper and originates in the lack of information for the derive_behaviour macro when the event type is specified. 
When the event type is not specified (it is generated) we can add the `From<Void>` conversion using a generics definition provided for the `Behaviour` struct with the `derive` attribute it seems we can rely on those generics to fill the `impl<..>` part of the conversion, on the other hand, when we specify `#[behaviour(to_swarm = "FooEvent")]` we are not able to get information about generics of this type. We probably could add properties of this attribute like (generics = None | SameAsBehaviour) (`#[behaviour(to_swarm = "FooEvent", no_generics)]`) but I'm not sure it is worth it.  @thomaseizinger Maybe we can discuss it during the next rust maintainer's call.

**Reply by thomaseizinger** [reply]
I am wondering why you put the `#impl_generics` block on the `impl From` block in https://github.com/shamil-gadelshin/rust-libp2p/commit/d6bb609568855a79ce7ce801ddcc81134687d8a6?

Those are the generics of the behaviour, not the event. What we can do is parse what is in `to_swarm` as `ItemType` which gives us access to all generics. We can then use those to correctly emit an `impl From` for the specified type.

```rust
#[derive(NetworkBehaviour)]
    #[behaviour(to_swarm = "OutEvent<O>")]
    struct Behaviour<T: 'static + Send> {
        custom: InnerBehaviour<T>,
    }
```

Does that make sense?
